src/PubSubClient.cpp

```diff
@@ -174,12 +186,16 @@ boolean PubSubClient::connect(const char *id, const char *user, const char *pass
             write(MQTTCONNECT,buffer,length-5);
 
             lastInActivity = lastOutActivity = millis();
-
+			
             while (!_client->available()) {
                 unsigned long t = millis();
                 if (t-lastInActivity >= ((int32_t) MQTT_SOCKET_TIMEOUT*1000UL)) {
                     _state = MQTT_CONNECTION_TIMEOUT;
                     _client->stop();
+
+                    // Next time, Try to Switch Protocol if client not responding 
+                    protocolType = (protocolType==MQTT_VERSION_3_1_1) ? MQTT_VERSION_3_1 : MQTT_VERSION_3_1_1; 
```

Comment:
Need to put some more thought on if automatically switching is good or not. Typically, you will know (or discover the first time you test) whether the protocol you are using is supported by the broker you are pointed at. You are highly unlikely to need to change protocols at runtime after the initial setup.